README.md
[![Build Status](https://secure.travis-ci.org/davidski/sch-logstash.png)](http://travis-ci.org/davidski/sch-logstash)

# sch-logstash cookbook

# Requirements

# Usage

# Attributes

# Recipes

# Author

Author:: David F. Severski (<davidski@deadheaven.com>)


CHANGELOG.md
# Changelog

## 0.2.0

Update to use 0.11.x series of lusis\chef-logstash

## 0.1.0

Inital release

recipes/local_test.rb
#
# Cookbook Name:: sch-logstash
# Recipe:: local_test
#
# Copyright (C) 2014 David F. Severski
# 
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
# 
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

name = 'server'

#clear out the default config templates
#attributes = node['logstash']['instance'][name]
#node.normal['logstash']['instance'][name]['config_templates'].keys.each do |k|
#  node.normal['logstash']['instance'][name]['config_templates'].delete(k)
#end

#node.force_override['logstash']['instance']['server']['config_templates'] = {
#  'input_redis' => 'config/input_redis.conf.erb',
#  'filter_sidewinder' => 'config/filter_sidewinder.conf.erb',
#  'output_elasticsearch' => 'config/output_elasticsearch.conf.erb'
#}

# set the number of workers threads to the number of cpus
node.normal['logstash']['instance'][name]['workers'] = node['cpu']['total']

# fetch our licensed version of MaxMind city DB
src_url = "https://download.maxmind.com/app/geoip_download?edition_id=133&suffix=tar.gz&license_key=#{node['logstash']['instance'][name]['maxmind_license_key']}"
src_filename = "GeoIPCity.tar.gz"
src_filepath = "#{Chef::Config['file_cache_path']}/#{src_filename}"
extract_path = "/opt/logstash/#{name}/vendor/geoip/MaxMind"

remote_file src_filepath do
  source src_url
  owner 'root'
  group 'root'
  mode 00644
  action :create_if_missing
end

bash 'extract_module' do
  cwd ::File.dirname(src_filepath)
  code <<-EOH
    mkdir -p #{extract_path}
    tar xzf #{src_filename} -C #{extract_path}
    chown -R logstash:logstash #{extract_path}
    mv #{extract_path}/*/* #{extract_path}/..
    EOH
  not_if { ::File.exists?(extract_path) }
end

# create the server instance
logstash_instance name do
  action            :create
end

es_ip = ::Logstash.service_ip(node, name, 'elasticsearch')
#   'output_stdout' => 'config/output_stdout.conf.erb'
my_config_templates = {
    'input_file' => 'config/input_file.conf.erb',
    'filter_sidewinder' => 'config/filter_sidewinder.conf.erb',
    'filter_webwasher' => 'config/filter_webwasher.conf.erb',
    'filter_metrics' => 'config/filter_metrics.conf.erb',
    'output_file' => 'config/output_file.conf.erb',
    'output_elasticsearch' => 'config/output_elasticsearch.conf.erb'
} 

# create our configuration files from the provided templates
logstash_config name do
  Chef::Log.debug("config vars: #{node['logstash']['instance'].inspect}")
  templates my_config_templates
  action [:create]
  variables(
      elasticsearch_ip: 			node['logstash']['instance']['elasticsearch_ip'],
      elasticsearch_embedded: false,
      elasticsearch_template: "/opt/logstash/#{name}/etc/elasticsearch_template.json",
      elasticsearch_cluster: 	node['logstash']['instance']['elasticsearch_cluster'],
      elasticsearch_protocol: "http",
      elasticsearch_workers:	node['cpu']['total'],
      logstash_host:					node['hostname'],
      input_redis_host: 			"10.0.0.21",
      input_redis_datatype: 	"list",
      input_redis_type: 			"sidewinder",
      output_graphite_host: 	"10.0.0.51",
      redis_workers:					1,
      input_file_exclude:			"*.gz",
      input_file_name:				"/vagrant_data/access*",
      input_file_type:				"webwasher",
      input_file_position:		"beginning",
      geoip_database:					"/opt/logstash/#{name}/vendor/geoip/GeoIPCity.dat"

  )
end

# create custom ES index template
template "/opt/logstash/server/etc/elasticsearch_template.json" do
  source "elasticsearch_template.json.erb"
  owner "logstash"
  group "logstash"
end

# create our custom patterns
logstash_pattern name do
  action [:create]
end


# enable and start the service
logstash_service name do
  action      [:enable, :start]
end


recipes/reader.rb
#
# Cookbook Name:: sch-logstash
# Recipe:: reader
#
# Copyright (C) 2014 David F. Severski
# 
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
# 
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

name = 'server'

#clear out the default config templates
#attributes = node['logstash']['instance'][name]
#node.normal['logstash']['instance'][name]['config_templates'].keys.each do |k|
#  node.normal['logstash']['instance'][name]['config_templates'].delete(k)
#end

# set the templates we want to use, for either cloud or local mode

# create the server instance
logstash_instance name do
  action            :create
end

# enable and start the service
logstash_service name do
  action      [:enable, :start]
end

if node.attribute?('cloud') && node['cloud']['provider'] == "ec2"
  my_config_templates = {
    'input_s3' => 'config/input_s3.conf.erb',
    'filter_metrics' => 'config/filter_metrics.conf.erb',
    'output_redis' => 'config/output_redis.conf.erb',
    'output_graphite' => 'config/output_graphite.conf.erb'
  }
else
  my_config_templates = {
    'input_file' => 'config/input_file.conf.erb',
    'filter_metrics' => 'config/filter_metrics.conf.erb',
    'output_redis' => 'config/output_redis.conf.erb',
    'output_graphite' => 'config/output_graphite.conf.erb'
  }
end

es_ip = service_ip(name, 'elasticsearch')

# create our configuration files from the provided templates
logstash_config name do
  Chef::Log.debug("config vars: #{node['logstash']['instance']['server'].inspect}")
  templates my_config_templates
  action [:create]
  variables(
    elasticsearch_ip: 								es_ip,
    elasticsearch_embedded: 					false,
    input_file_exclude: 							"*.gz",
    input_file_name: 									"/vagrant_data/13*",
    input_file_type: 									"sidewinder",
    input_file_position: 							"beginning",
    input_s3_bucket: 									"log-inbox.elk.sch",
    input_s3_delete_after_read: 			true,
    input_s3_region: 									"us-west-2",
    input_s3_bucket_access_key_id: 		node['logstash']['instance']['server']['aws_access_key_id'],
    input_s3_bucket_secret_access_key: node['logstash']['instance']['server']['aws_secret_access_key'],
    output_redis_datatype: 						"list",
    output_redis_host: 								"10.0.0.21",
    output_graphite_host: 						"10.0.0.51"
  )
end

# create our custom patterns
logstash_pattern name do
  action [:create]
end

recipes/server.rb
#
# Cookbook Name:: sch-logstash
# Recipe:: server
#
# Copyright (C) 2014 David F. Severski
# 
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
# 
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

name = 'server'

# create the server instance
logstash_instance name do
  action            :create
end

# enable and start the service
logstash_service name do
  action      [:enable, :start]
end

es_ip = service_ip(name, 'elasticsearch')

# create our configuration files from the provided templates
logstash_config name do
  Chef::Log.info("config vars: #{node['logstash']['instance']['server'].inspect}")
  action [:create]
  variables(
      elasticsearch_ip: es_ip,
      elasticsearch_embedded: false
  )
end

# create our custom patterns
logstash_pattern name do
  action [:create]
end

recipes/default.rb
#
# Cookbook Name:: sch-logstash
# Recipe:: default
#
# Copyright (C) 2014 David F. Severski
# 
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
# 
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
include_recipe "logstash"
begin
	r = resources(:template => "#{node['logstash']['basedir']}/templates/elk_test.conf.erb")
	r.cookbooks "sch-logstash"
rescue Chef::Exceptions::ResourceNotFound
	Chef::Log.warn "Could not find template to override!"
end



recipes/shipper.rb
#
# Cookbook Name:: sch-logstash
# Recipe:: shipper
#
# Copyright (C) 2014 David F. Severski
# 
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
# 
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

include_recipe 'chef-sugar::default'

name = 'server'

#clear out the default config templates
#attributes = node['logstash']['instance'][name]
#node.normal['logstash']['instance'][name]['config_templates'].keys.each do |k|
#  node.normal['logstash']['instance'][name]['config_templates'].delete(k)
#end

# set the number of workers threads to the number of cpus
node.normal['logstash']['instance'][name]['workers'] = node['cpu']['total']

# create the server instance
logstash_instance name do
  action            :create
  #workers           node['cpu']['total']
end

# enable and start the service
logstash_service name do
  action      [:enable]
end

# set the templates we want to use, for either cloud or local mode
if node.attribute?('cloud') && node['cloud']['provider'] == "ec2"
  my_config_templates = {
    'input_s3' => 'config/input_s3.conf.erb',
    'filter_metrics' => 'config/filter_metrics.conf.erb',
    'output_redis' => 'config/output_redis.conf.erb',
    'output_graphite' => 'config/output_graphite.conf.erb'
  }
else
  my_config_templates = {
    'input_file' => 'config/input_file.conf.erb',
    'filter_metrics' => 'config/filter_metrics.conf.erb',
    'output_redis' => 'config/output_redis.conf.erb',
    'output_graphite' => 'config/output_graphite.conf.erb'
  }
end

# discovery for redis ip
if Chef::Config[:solo]
    redis_ip = "10.0.0.21"
else
    #redis = search('node', 'role:redis').first
    #redis_ip = best_ip_for(redis)
    redis = node['opsworks']['layers']['redis']['instances'].first[1]
    redis_ip = redis['private_ip']
end

# discovery for graphite IP
if Chef::Config[:solo]
    graphite_ip = "10.0.0.50"
else
    #kb = search('node', 'role:kb').first
    #graphite_ip = best_ip_for(kb)
    kb = node['opsworks']['layers']['kb']['instances'].first[1]
    Chef::Log.warn("OpsWorks Kibana layer reads: #{kb}")
    graphite_ip = kb['private_ip']
end

# create our configuration files from the provided templates
logstash_config name do
  Chef::Log.debug("config vars: #{node['logstash']['instance']['server'].inspect}")
  templates my_config_templates
  action [:create]
  variables(
    elasticsearch_ip: 								::Logstash.service_ip(node, name, 'elasticsearch'),
    elasticsearch_embedded: 					false,
    logstash_host:										node['hostname'],
    input_file_exclude: 							"*.gz",
    input_file_name: 									"/vagrant_data/13*",
    input_file_type: 									"sidewinder",
    input_file_position: 							"beginning",
    input_s3_bucket: 									"log-inbox.elk.sch",
    input_s3_prefix: 									"2-ELK/input",
    input_s3_delete_after_read: 			true,
    input_s3_region: 									"us-west-2",
    input_s3_bucket_access_key_id: 		node['logstash']['instance']['server']['aws_access_key_id'],
    input_s3_bucket_secret_access_key: node['logstash']['instance']['server']['aws_secret_access_key'],
    output_redis_datatype: 						"list",
    output_redis_host: 								redis_ip,
    redis_batch_size:									"1000",
    output_graphite_host: 						graphite_ip,
    redis_workers:										node['cpu']['total'],
    redis_congestion_threshold:				20000000,
    redis_congestion_interval:				60
  )
  notifies :restart, "logstash_service[#{name}]"
end

logstash_plugins "contrib" do
  instance name
  action [:create]
end

# create our custom patterns
logstash_pattern name do
  action [:create]
end

recipes/indexer.rb
#
# Cookbook Name:: sch-logstash
# Recipe:: indexer
#
# Copyright (C) 2014 David F. Severski
# 
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
# 
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

include_recipe 'chef-sugar::default'

name = 'server'

#clear out the default config templates
#attributes = node['logstash']['instance'][name]
#node.normal['logstash']['instance'][name]['config_templates'].keys.each do |k|
#  node.normal['logstash']['instance'][name]['config_templates'].delete(k)
#end

#node.force_override['logstash']['instance']['server']['config_templates'] = {
#  'input_redis' => 'config/input_redis.conf.erb',
#  'filter_sidewinder' => 'config/filter_sidewinder.conf.erb',
#  'output_elasticsearch' => 'config/output_elasticsearch.conf.erb'
#}

# set the number of workers threads to the number of cpus
node.normal['logstash']['instance'][name]['workers'] = node['cpu']['total']

# fetch our licensed version of MaxMind city DB
src_url = "https://download.maxmind.com/app/geoip_download?edition_id=133&suffix=tar.gz&license_key=#{node['logstash']['instance'][name]['maxmind_license_key']}"
src_filename = "GeoIPCity.tar.gz"
src_filepath = "#{Chef::Config['file_cache_path']}/#{src_filename}"
extract_path = "/opt/logstash/#{name}/vendor/geoip/MaxMind"

remote_file src_filepath do
  source src_url
  owner 'root'
  group 'root'
  mode 00644
end

bash 'extract_module' do
  cwd ::File.dirname(src_filepath)
  code <<-EOH
    mkdir -p #{extract_path}
    tar xzf #{src_filename} -C #{extract_path}
    chown -R logstash:logstash #{extract_path}
    mv #{extract_path}/*/* #{extract_path}/..
    EOH
  not_if { ::File.exists?(extract_path) }
end

# create the server instance
logstash_instance name do
  action            :create
  #workers           node['cpu']['total']
end

es_ip = ::Logstash.service_ip(node, name, 'elasticsearch')
my_config_templates = {
    'input_redis' => 'config/input_redis.conf.erb',
    'filter_sidewinder' => 'config/filter_sidewinder.conf.erb',
    'filter_metrics' => 'config/filter_metrics.conf.erb',
    'output_elasticsearch' => 'config/output_elasticsearch.conf.erb',
    'output_graphite' => 'config/output_graphite.conf.erb'
}

# discovery for redis IP
if Chef::Config[:solo]
    redis_ip = "10.0.0.21"
else
    #redis = search('node', 'role:redis').first
    #redis_ip = best_ip_for(redis)
    redis = node['opsworks']['layers']['redis']['instances'].first[1]
    Chef::Log.warn("OpsWorks Redis layer reads: #{redis}")
    redis_ip = redis['private_ip']
end

# discovery for graphite IP
if Chef::Config[:solo]
    graphite_ip = "10.0.0.50"
else
    #kb = search('node', 'role:kb').first
    #graphite_ip = best_ip_for(kb)
    kb = node['opsworks']['layers']['kb']['instances'].first[1]
    Chef::Log.warn("OpsWorks Kibana layer reads: #{kb}")
    graphite_ip = kb['private_ip']
end

# create our configuration files from the provided templates
logstash_config name do
  Chef::Log.debug("config vars: #{node['logstash']['instance'].inspect}")
  templates my_config_templates
  action [:create]
  variables(
      elasticsearch_ip: 			"internal-elasticsearch-lb-532683521.us-west-2.elb.amazonaws.com",
      elasticsearch_embedded: false,
      elasticsearch_template: "/opt/logstash/#{name}/etc/elasticsearch_template.json",
      elasticsearch_cluster: "elkrun",
      elasticsearch_protocol: "http",
      elasticsearch_workers:	node['cpu']['total'],
      logstash_host:					node['hostname'],
      input_redis_host: 			redis_ip,
      input_redis_datatype: 	"list",
      input_redis_type: 			"sidewinder",
      output_graphite_host: 	graphite_ip,
      redis_workers:					1,
      geoip_database:					"/opt/logstash/#{name}/vendor/geoip/GeoIPCity.dat"

  )
  notifies :restart, "logstash_service[#{name}]"
end

# create custom ES index template
template "/opt/logstash/server/etc/elasticsearch_template.json" do
  source "elasticsearch_template.json.erb"
  owner "logstash"
  group "logstash"
end

logstash_plugins "contrib" do
  instance name
  action [:create]
end

# create our custom patterns
logstash_pattern name do
  action [:create]
end

# enable the LS service
logstash_service name do
  action            [:enable]
end
